File: python/leetcode/305.py
class Solution(object):
    def numIslands2(self, m, n, positions):
        cnts, res, roots = 0, [], [-1] * (m * n)
        dp = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        for (x, y) in positions:
            idx = x * n + y
            roots[idx] = idx
            cnts += 1
            for (dx, dy) in dp:
                i, j = x + dx, y + dy
                pos = i * n + j
                if i >= 0 and i < m and j >= 0 and j < n and roots[pos] != -1:
                    rootIdx, rootPos = idx, pos
                    while rootIdx != roots[rootIdx]:
                        rootIdx = roots[rootIdx]
                    while rootPos != roots[rootPos]:
                        rootPos = roots[rootPos]
                    if rootIdx != rootPos:
                        roots[rootIdx] = rootPos
                        cnts -= 1
            res.append(cnts)
        return res

File: cpp/leetcode/305.cpp
class Solution {
public:
    vector<int> numIslands2(int m, int n, vector<pair<int, int>>& positions) {
        vector<int> res;
        vector<int> root(m * n, -1);
        vector<pair<int, int>> dp = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int island = 0;
        for (auto it : positions) {
            int x = it.first, y = it.second, idx = x * n + y; // idx is the new point
            root[idx] = idx;
            ++island;
            for (auto p : dp) {
                int i = x + p.first, j = y + p.second, pos = i * n + j;
                if (i >= 0 && i < m && j >= 0 && j < n && root[pos] != -1) {
                    int rootPos = pos, rootIdx = idx;
                    while(rootPos != root[rootPos]) rootPos = root[rootPos];
                    while(rootIdx != root[rootIdx]) rootIdx = root[rootIdx];
                    if (rootIdx != rootPos) {
                        root[rootIdx] = rootPos;
                        --island;
                    }
                }
            }
            res.push_back(island);
        }
        return res;
    }
};